File: TDDMicroExercises/Ruby/telemetry_system/telemetry_client.rb
class TelemetryClient
  DIAGNOSTICMESSAGE = "AT#UD"

  def initialize
    @online_status = false
    @diagnostic_message_result = ""
  end

  def online_status
    @online_status
  end

  def connect(telemetry_server_connection_string)
    if (telemetry_server_connection_string.nil? || telemetry_server_connection_string == "")
      raise Exception, "Argument Null"
    end

    #simulate the operation on a real modem
    success = (1 + rand(10)) <= 8

    @online_status = success
  end

  def disconnect 
    @online_status = false
  end

  def send(message)
    if (message.nil? || message == "")
      raise Exception, "Argument Null"
    end

    if (message == DIAGNOSTICMESSAGE)
      #simulate a status report
      @diagnostic_message_result =
        "LAST TX rate................ 100 MBPS\nHIGHEST TX rate............. 100 MBPS\nLAST RX rate................ 100 MBPS\nHIGHEST RX rate............. 100 MBPS\nBIT RATE.................... 100000000\nWORD LEN.................... 16\nWORD/FRAME.................. 511\nBITS/FRAME.................. 8192\nMODULATION TYPE............. PCM/FM\nTX Digital Los.............. 0.75\nRX Digital Los.............. 0.10\nBEP Test.................... -5\nLocal Rtrn Count............ 00\nRemote Rtrn Count........... 00"

      return
    end

    #here should go the real Send operation
  end

  def receive 
    if (!@diagnostic_message_result.nil? && @diagnostic_message_result != "")
      message = @diagnostic_message_result
      @diagnostic_message_result = "" 
    else
      #simulate a received message
      message = "" 
      message_length = 50 + rand(110)
      i = message_length 
      while i >= 0
        message += (40 + rand(126)).chr
        i -= 1
      end
    end

    return message
  end
end

File: TDDMicroExercises/PHP/TelemetrySystem/TelemetryClient.php
<?php

namespace TDDMicroExercises\PHP\TelemetrySystem;

class TelemetryClient
{
	const DiagnosticMessage = "AT#UD";

	private $_onlineStatus;
	private $_diagnosticMessageJustSent = false;

	/*private $_connectionEventsSimulator = new Random();
    private $_randomMessageSimulator 	= new Random();*/

    public function getOnlineStatus()
    {
    	return $this->_onlineStatus;
    }

    public function connect( $telemetryServerConnectionString )
    {
    	if( is_null($telemetryServerConnectionString) || empty($telemetryServerConnectionString) ) 
    	{
    		throw new \InvalidArgumentException();
    	}

    	// Fake the connection with 20% chances of success
		//$success = rand(1, 10) <= 2;
		$success = rand(1, 10) <= 2;
		$this->_onlineStatus = $success;
    }

    public function disconnect()
    {
    	$this->_onlineStatus = false;
    }

    public function send( $message )
    {
    	if( is_null($message) || empty($message) ) {
    		throw new \InvalidArgumentException();
    	}

    	// The simulation of Send() actually just remember if the last message sent was a diagnostic message.
        // This information will be used to simulate the Receive(). Indeed there is no real server listening.
        if ( $message == TelemetryClient::DiagnosticMessage )
		{
		    $this->_diagnosticMessageJustSent = true;

		}
		else
		{
		    $this->_diagnosticMessageJustSent = false;
		}
	
    }

    public function receive()
    {
    	$message = '';

    	if($this->_diagnosticMessageJustSent)
    	{
    		// Simulate the reception of the diagnostic message
                $message = "LAST TX rate................ 100 MBPS\r\n"
                    . "HIGHEST TX rate............. 100 MBPS\r\n"
                    . "LAST RX rate................ 100 MBPS\r\n"
                    . "HIGHEST RX rate............. 100 MBPS\r\n"
                    . "BIT RATE.................... 100000000\r\n"
                    . "WORD LEN.................... 16\r\n"
                    . "WORD/FRAME.................. 511\r\n"
                    . "BITS/FRAME.................. 8192\r\n"
                    . "MODULATION TYPE............. PCM/FM\r\n"
                    . "TX Digital Los.............. 0.75\r\n"
                    . "RX Digital Los.............. 0.10\r\n"
                    . "BEP Test.................... -5\r\n"
                    . "Local Rtrn Count............ 00\r\n"
                    . "Remote Rtrn Count........... 00";

                $this->_diagnosticMessageJustSent = false;
    	}
    	else
    	{
    		$messageLength = rand( 50, 110 );
    		for($i = $messageLength; $i > 0; $i-- )
    		{
    			$message += chr( rand( 40, 126 ) ) ;
    		}
    	}

    	return $message;
    }
}